app/src/main/java/org/coepi/android/ble/covidwatch/BLEAdvertiser.kt

```diff
@@ -0,0 +1,255 @@
+
+                try {
+                    if (characteristic?.uuid == UUIDs.CONTACT_EVENT_IDENTIFIER_CHARACTERISTIC) {
+                        if (offset != 0) {
+                            result = BluetoothGatt.GATT_INVALID_OFFSET
+                            return
+                        }
+
+                        val newContactEventIdentifier = UUID.randomUUID()
+
+                        logContactEventIdentifier(newContactEventIdentifier)
+
+                        value = newContactEventIdentifier.toBytes()
+                    } else {
+                        result = BluetoothGatt.GATT_FAILURE
+                    }
+                } catch (exception: Exception) {
+                    result = BluetoothGatt.GATT_FAILURE
+                    value = null
+                } finally {
+                    Log.i(
+                        TAG,
+                        "onCharacteristicReadRequest result=$result device=$device requestId=$requestId offset=$offset characteristic=$characteristic"
+                    )
+                    bluetoothGattServer?.sendResponse(
+                        device,
+                        requestId,
+                        result,
+                        offset,
+                        value
+                    )
+                }
+            }
+
+            override fun onCharacteristicWriteRequest(
+                device: BluetoothDevice?,
+                requestId: Int,
+                characteristic: BluetoothGattCharacteristic?,
+                preparedWrite: Boolean,
+                responseNeeded: Boolean,
+                offset: Int,
+                value: ByteArray?
+            ) {
+                super.onCharacteristicWriteRequest(
+                    device,
+                    requestId,
+                    characteristic,
+                    preparedWrite,
+                    responseNeeded,
+                    offset,
+                    value
+                )
+
+                var result = BluetoothGatt.GATT_SUCCESS
+                try {
+                    if (characteristic?.uuid == UUIDs.CONTACT_EVENT_IDENTIFIER_CHARACTERISTIC) {
+                        if (offset != 0) {
+                            result = BluetoothGatt.GATT_INVALID_OFFSET
+                            return
+                        }
+
+                        val newContactEventIdentifier = value?.toUUID()
+                        if (newContactEventIdentifier == null) {
+                            result = BluetoothGatt.GATT_FAILURE
+                            return
+                        }
+
+                        logContactEventIdentifier(newContactEventIdentifier)
+                    } else {
+                        result = BluetoothGatt.GATT_FAILURE
+                    }
+                } catch (exception: Exception) {
+                    result = BluetoothGatt.GATT_FAILURE
+                } finally {
+                    Log.i(
```

Comment:
hopefully will have connection here to CenManager

Reply:
The idea is to decouple BLE from our CEN functionality. The advertiser just emits events and `CenManager` observers from "outside", similar to the scanner, with `scanObservable`.

Reply:
I can trigger a callback / emit an event here, that will be processed by `CenManager `, like it's done in `BLEScanner`.

Reply:
Addressed. Now the write is handled and broadcasted as a scan event (since it's a workaround for it AFAIU).